// amazeui-touch/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Container = require('./Container');

Object.defineProperty(exports, 'Container', {
  enumerable: true,
  get: function get() {
    return _Container.default;
  }
});

var _Grid = require('./Grid');

Object.defineProperty(exports, 'Grid', {
  enumerable: true,
  get: function get() {
    return _Grid.default;
  }
});

var _Col = require('./Col');

Object.defineProperty(exports, 'Col', {
  enumerable: true,
  get: function get() {
    return _Col.default;
  }
});

var _Group = require('./Group');

Object.defineProperty(exports, 'Group', {
  enumerable: true,
  get: function get() {
    return _Group.default;
  }
});

var _Accordion = require('./Accordion');

Object.defineProperty(exports, 'Accordion', {
  enumerable: true,
  get: function get() {
    return _Accordion.default;
  }
});

var _Badge = require('./Badge');

Object.defineProperty(exports, 'Badge', {
  enumerable: true,
  get: function get() {
    return _Badge.default;
  }
});

var _Button = require('./Button');

Object.defineProperty(exports, 'Button', {
  enumerable: true,
  get: function get() {
    return _Button.default;
  }
});

var _ButtonGroup = require('./ButtonGroup');

Object.defineProperty(exports, 'ButtonGroup', {
  enumerable: true,
  get: function get() {
    return _ButtonGroup.default;
  }
});

var _ButtonToolbar = require('./ButtonToolbar');

Object.defineProperty(exports, 'ButtonToolbar', {
  enumerable: true,
  get: function get() {
    return _ButtonToolbar.default;
  }
});

var _Card = require('./Card');

Object.defineProperty(exports, 'Card', {
  enumerable: true,
  get: function get() {
    return _Card.default;
  }
});

var _Icon = require('./Icon');

Object.defineProperty(exports, 'Icon', {
  enumerable: true,
  get: function get() {
    return _Icon.default;
  }
});

var _Field = require('./Field');

Object.defineProperty(exports, 'Field', {
  enumerable: true,
  get: function get() {
    return _Field.default;
  }
});

var _List = require('./List');

Object.defineProperty(exports, 'List', {
  enumerable: true,
  get: function get() {
    return _List.default;
  }
});

var _Loader = require('./Loader');

Object.defineProperty(exports, 'Loader', {
  enumerable: true,
  get: function get() {
    return _Loader.default;
  }
});

var _Modal = require('./Modal');

Object.defineProperty(exports, 'Modal', {
  enumerable: true,
  get: function get() {
    return _Modal.default;
  }
});

var _NavBar = require('./NavBar');

Object.defineProperty(exports, 'NavBar', {
  enumerable: true,
  get: function get() {
    return _NavBar.default;
  }
});

var _Notification = require('./Notification');

Object.defineProperty(exports, 'Notification', {
  enumerable: true,
  get: function get() {
    return _Notification.default;
  }
});

var _OffCanvas = require('./OffCanvas');

Object.defineProperty(exports, 'OffCanvas', {
  enumerable: true,
  get: function get() {
    return _OffCanvas.default;
  }
});

var _OffCanvasTrigger = require('./OffCanvasTrigger');

Object.defineProperty(exports, 'OffCanvasTrigger', {
  enumerable: true,
  get: function get() {
    return _OffCanvasTrigger.default;
  }
});

var _Popover = require('./Popover');

Object.defineProperty(exports, 'Popover', {
  enumerable: true,
  get: function get() {
    return _Popover.default;
  }
});

var _PopoverTrigger = require('./PopoverTrigger');

Object.defineProperty(exports, 'PopoverTrigger', {
  enumerable: true,
  get: function get() {
    return _PopoverTrigger.default;
  }
});

var _Slider = require('./Slider');

Object.defineProperty(exports, 'Slider', {
  enumerable: true,
  get: function get() {
    return _Slider.default;
  }
});

var _Switch = require('./Switch');

Object.defineProperty(exports, 'Switch', {
  enumerable: true,
  get: function get() {
    return _Switch.default;
  }
});

var _TabBar = require('./TabBar');

Object.defineProperty(exports, 'TabBar', {
  enumerable: true,
  get: function get() {
    return _TabBar.default;
  }
});

var _Tabs = require('./Tabs');

Object.defineProperty(exports, 'Tabs', {
  enumerable: true,
  get: function get() {
    return _Tabs.default;
  }
});

var _Touchable = require('./Touchable');

Object.defineProperty(exports, 'Touchable', {
  enumerable: true,
  get: function get() {
    return _Touchable.default;
  }
});

var _View = require('./View');

Object.defineProperty(exports, 'View', {
  enumerable: true,
  get: function get() {
    return _View.default;
  }
});

var _mixins = require('./mixins');

var _loop = function _loop(_key2) {
  if (_key2 === "default") return 'continue';
  Object.defineProperty(exports, _key2, {
    enumerable: true,
    get: function get() {
      return _mixins[_key2];
    }
  });
};

for (var _key2 in _mixins) {
  var _ret = _loop(_key2);

  if (_ret === 'continue') continue;
}

// TODO: why `export Button from './Button'` not works?
// @see http://jamesknelson.com/re-exporting-es6-modules/
// @see https://github.com/Microsoft/TypeScript/issues/2726

var VERSION = exports.VERSION = '0.1.0-beta2';

// Layout

// Mixins
